### ISSUE TITLE
model.fit() :  inconsistant behaviour with `max_fev`  + working with non-existing method ??

### ISSUE THREAD

**Issue by patquem** [lmfit/lmfit-py#842]
Hi ,
You worked a while on `max_fev` (#610 ), and some things related to that argument are still unclear to me.
I  created the example below which illustrates some strange and confusing behaviour for me.
Here are some comments/questions that this example inspire me:
1 -  using `max_nfev` works fine in case of `leastsq` when `max_nfev` is reached
2 - using `leasts_squares` : a/ `nfev` is abnormally low compared to those obtained with `leastsq` b/ when `max_nfev` seems to be reached, `nfev` radically changes of value (to be more conform with the "real" and expected `nfev` I think)
3 - `least_square` (without an 's' - that is not referenced) returns the same results than 'least_squares'
4 - A non-existing fitting method ('xxxxx') returns nevertheless a result (??? from which algorithm ???)
Thank you for your help,
Patrick

lmfit version : 1.0.3

```
import numpy as np
import matplotlib.pyplot as plt
from lmfit import Model
from lmfit.models import LorentzianModel

x = np.arange(200)
y = np.sin(0.2 * x).clip(0)


def eval(method, max_nfev=None):
    label = f"{method} - max_nfev={max_nfev}"
    print(label)
    plt.figure()
    plt.title(label)
    plt.plot(x, y)
    for i in range(4):
        model = LorentzianModel(prefix=f'm{i}_')
        model.set_param_hint("amplitude", min=0)
        model.set_param_hint("sigma", max=10)
        model.set_param_hint("center", value=40 * (i + 1))
        if i == 0:
            models = model
        else:
            models += model
        out = models.fit(y, x=x, method=method, max_nfev=max_nfev)
        print(i, out.nfev)
        plt.plot(x, out.best_fit, label=f'fit-{i + 1}')
    plt.legend()
    print()


eval(method='leastsq', max_nfev=None)
eval(method='leastsq', max_nfev=150)
eval(method='least_squares', max_nfev=None)
eval(method='least_squares', max_nfev=150)

eval(method='least_square', max_nfev=None)
eval(method='xxxxx', max_nfev=None)

plt.show()
```
```
leastsq - max_nfev=None
0 61
1 73
2 133
3 212

leastsq - max_nfev=150
0 61
1 73
2 133
3 150

least_squares - max_nfev=None
0 11
1 11
2 17
3 14

least_squares - max_nfev=150
0 11
1 11
2 151
3 151

least_square - max_nfev=None
0 11
1 11
2 17
3 14

xxxxx - max_nfev=None
0 170
1 1513
2 2787
3 5366
```

**Comment by newville** [comment]
@patquem Thanks - that's interesting and worth figuring out in more detail.  I think that `least_sq` will be OK to specify the `method` is not found, and that if the method is not found, a scalar minimizer will be used.

But also: I believe that least_squares is just lying about the number of function evaluations.  I don't have a lot of love for that method ;).

I think we are not going to get hung up on +/- a few calls that exceed max_nfev.

**Comment by patquem** [comment]
Hi,
Looking at the lmfit src more in details:
3/ `least_square` returns same results than `least_squares` because case is defined with only the first letters `least_sq` (strange and not straightforward, but I can assume there was a reason for that)
4/ 'xxxxx' runs with `scalar_minimize(method='Nelder-Mead')` as suggested by @newville . On my opinion, a non-existing model should return an error message, don't you ?
1/-2/ a consistent `nfev `for `least_squares `would be awesome :)
Patrick

**Comment by newville** [comment]
@patquem 

Well, having "least_square" work for "least_squares" is mostly just to better support sort of obvious mis-spellings. 
Having "xxxxx" mean "Nelder-Mead" is probably just an extension of that. We could require correctly spelled names, but then we have to be willing to maintain that list of allowed methods.  I suppose we could do that...

On `scipy.optimize.least_squares`, let me just say that this code is utterly contemptable.  The API for this method (yes, for this method alone) is a master class in "Don't do that".  And, yep, it does report an artificially low value for `nfev`.  The docstring for this says:
            nfev : int
                Number of function evaluations done. Methods 'trf' and 'dogbox' do
                not count function calls for numerical Jacobian approximation, as
                opposed to 'lm' method.
         
That is to say, that it deliberately and purposefully excludes calls to the objective function when calculating derivatives using the default `trf` and `dogbox` method,  but not when using `lm` (which, as far as I can tell is exactly `leastsq`).    The returned value for `nfev` not only has a different meaning from other methods in `scipy.optimize`, but it is not even internally consistent.   Yikes. 

It dramatically undercounts function calls (like by a factor of 3 !) on purpose, knowing that it will be compared to other optimization methods within `scipy.optimize`.  In fact, `trf` typically actually uses more function evaluations than `leastsq` does.  It typically reports that it uses far fewer.  

It is, in short, lying.  The deception is deliberate.  They are purposefully presenting their users with information they know is not correct.

A branch I am working on at https://github.com/lmfit/lmfit-py/tree/better_make_params  (which is mostly about making it easier to create parameters objects) will change this to report the real number as `nfev` and their deliberately deceptive value as `least_squares_nfev`.

I will say, that we don't count every function call either.  We count the number of function calls made **during the solving process**.   There are typically a few additional function calls made (between 2 and 5). If we need to calculate uncertainties after a fit that has not generated these already, that will involve roughly 2*(n_varys+1) more actual function calls. 

It is sort of worth considering why we are supporting the use of `least_squares`.

**Comment by newville** [comment]
@patquem This is addressed in 1.2.0, so closing here.
